launcher/FileSystem.cpp

```diff
@@ -215,6 +302,271 @@ bool copy::operator()(const QString& offset, bool dryRun)
+                auto relative_path = src_dir.relativeFilePath(src_path);
+
+                if (m_max_depth >= 0 && PathDepth(relative_path) > m_max_depth) {
+                    relative_path = PathTruncate(relative_path, m_max_depth);
+                    src_path = src_dir.filePath(relative_path);
+                    if (linkedPaths.contains(src_path)) {
+                        continue;
+                    }
+                }
+
+                linkedPaths.append(src_path);
+
+                link_file(src_path, relative_path);
+            }
+        }
+    }
+}
+
+bool create_link::make_links()
+{
+    for (auto link : m_links_to_make) {
+        QString src_path = link.src;
+        QString dst_path = link.dst;
+
+        ensureFilePathExists(dst_path);
+        if (m_useHardLinks) {
+            if (m_debug)
+                qDebug() << "making hard link:" << src_path << "to" << dst_path;
+            fs::create_hard_link(StringUtils::toStdString(src_path), StringUtils::toStdString(dst_path), m_os_err);
+        } else if (fs::is_directory(StringUtils::toStdString(src_path))) {
+            if (m_debug)
+                qDebug() << "making directory_symlink:" << src_path << "to" << dst_path;
+            fs::create_directory_symlink(StringUtils::toStdString(src_path), StringUtils::toStdString(dst_path), m_os_err);
+        } else {
+            if (m_debug)
+                qDebug() << "making symlink:" << src_path << "to" << dst_path;
+            fs::create_symlink(StringUtils::toStdString(src_path), StringUtils::toStdString(dst_path), m_os_err);
+        }
+
+        if (m_os_err) {
+            qWarning() << "Failed to link files:" << QString::fromStdString(m_os_err.message());
+            qDebug() << "Source file:" << src_path;
+            qDebug() << "Destination file:" << dst_path;
+            qDebug() << "Error category:" << m_os_err.category().name();
+            qDebug() << "Error code:" << m_os_err.value();
+            emit linkFailed(src_path, dst_path, QString::fromStdString(m_os_err.message()), m_os_err.value());
+        } else {
+            m_linked++;
+            emit fileLinked(src_path, dst_path);
+        }
+        if (m_os_err)
+            return false;
+    }
+    return true;
+}
+
+void create_link::runPrivileged(const QString& offset)
+{
+    m_linked = 0;  // reset counter
+    m_path_results.clear();
+    m_links_to_make.clear();
+
+    bool gotResults = false;
+
+    make_link_list(offset);
+
+    QString serverName = BuildConfig.LAUNCHER_APP_BINARY_NAME + "_filelink_server" + StringUtils::getRandomAlphaNumeric();
+
+    connect(&m_linkServer, &QLocalServer::newConnection, this, [&]() {
+        qDebug() << "Client connected, sending out pairs";
+        // construct block of data to send
+        QByteArray block;
+        QDataStream out(&block, QIODevice::WriteOnly);
+        out.setVersion(QDataStream::Qt_5_0);  // choose correct version better?
```

Comment:
The types are being encoded as raw bytes and set over a socket in order for the under end to know how to reconstruct the types on the other end both sides have to agree on a format for the bytes to be in. each version of QT has new or changed types and so needs a new protocol version. by setting a version here Qt knows how to encode and decode the byte data.

Reply:
Yes, but both sides of the communication are using `QDataStream` in the very same Qt version. It may be a raw binary encoding, but `QDataStream` will always (de)serialize that data in the same way on both sides, so it shouldn't matter which exact version we're using, provided we'll use the same one as the default anyway.

Reply:
choosing a protocol version that is stable and leaving it seems like a good solution. this way it won't unexpectedly break if we have an update and for some reason one binary isn't updated correctly.

Reply:
> this way it won't unexpectedly break if we have an update and for some reason one binary isn't updated correctly

how would that even happen... I'm more worried about fixing the version and that specific protocol version being accidentally broken in a new Qt version without anyone noticing it, since there's so many of them. I would assume it would be harder for that to happen in the default version, since that's the most likely to be tested.

Reply:
I suppose you have a point. I was more worried that if there was an update and the main launcher updated by the link binary didn't it would break. but that is a fixable problem I guess.

I went ahead and removed the explicit version.